calyx/src/analysis/reaching_defns.rs

```diff
@@ -0,0 +1,422 @@
+use std::cmp::Ordering;
+use std::cmp::{Ord, PartialOrd};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    ops::BitOr,
+    rc::Rc,
+};
+
+use crate::analysis::ReadWriteSet;
+use crate::ir::{self, RRC};
+
+pub const INVOKE_PREFIX: &str = "__invoke_";
+
+type GroupName = ir::Id;
+type InvokeName = ir::Id;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub enum GroupOrInvoke {
+    Group(GroupName),
+    Invoke(InvokeName),
+}
+
+impl PartialOrd for GroupOrInvoke {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for GroupOrInvoke {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (self, other) {
+            (GroupOrInvoke::Group(a), GroupOrInvoke::Group(b))
+            | (GroupOrInvoke::Invoke(a), GroupOrInvoke::Invoke(b)) => {
+                ir::Id::cmp(a, b)
+            }
+            (GroupOrInvoke::Group(_), GroupOrInvoke::Invoke(_)) => {
+                Ordering::Greater
+            }
+            (GroupOrInvoke::Invoke(_), GroupOrInvoke::Group(_)) => {
+                Ordering::Less
+            }
+        }
+    }
+}
+
+impl Into<ir::Id> for GroupOrInvoke {
+    fn into(self) -> ir::Id {
+        match self {
+            GroupOrInvoke::Group(id) | GroupOrInvoke::Invoke(id) => id,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DefSet {
```

Comment:
Since this is a public struct, it needs a doc comment describing how this is constructed from the IR and how it can be used. I think the live range struct might have a comment that's a useful example of how to do it.